File: data/src/main/java/home/oleg/placesnearme/repositories/FavoriteVenueRepositoryImpl.java
package home.oleg.placesnearme.repositories;

import java.util.List;

import home.oleg.placenearme.models.DetailedVenue;
import home.oleg.placenearme.repositories.FavoriteVenuesRepository;
import home.oleg.placesnearme.dao.DetailedVenueWithPhotosDao;
import io.reactivex.Completable;
import io.reactivex.Single;

public class FavoriteVenueRepositoryImpl implements FavoriteVenuesRepository {

    private final DetailedVenueWithPhotosDao dao;

    public FavoriteVenueRepositoryImpl(DetailedVenueWithPhotosDao dao) {
        this.dao = dao;
    }

    @Override
    public Single<List<DetailedVenue>> getFavorites() {
        return null;
    }

    @Override
    public Completable addToFavorite(DetailedVenue venue) {
        return Completable.complete();
    }

    @Override
    public Completable deleteFromFavorite(String id) {
        return Completable.complete();
    }

}

File: core-data/src/main/java/home/oleg/placesnearme/coredata/repositories/FavoriteVenueRepositoryImpl.kt
package home.oleg.placesnearme.coredata.repositories

import home.oleg.placesnearme.coredata.dao.PlacesDao
import home.oleg.placesnearme.coredata.mapper.DetailedVenueMapper
import home.oleg.placesnearme.coredomain.models.Place
import home.oleg.placesnearme.coredomain.repositories.FavoriteVenuesRepository
import io.reactivex.Completable
import io.reactivex.Flowable
import javax.inject.Inject

class FavoriteVenueRepositoryImpl @Inject constructor(private val dao: PlacesDao) : FavoriteVenuesRepository {

    override fun observeFavorites(): Flowable<List<Place>> {
        return dao.streamFavorites().map { DetailedVenueMapper.map(it) }
    }

    override fun addToFavorite(venue: Place): Completable {
        return Completable.fromAction {
            val placeAndPhotos = DetailedVenueMapper.map(venue)
            placeAndPhotos.venue = placeAndPhotos.venue.copy(isFavorite = true)
            dao.insertOrReplace(placeAndPhotos.venue, placeAndPhotos.photos)
        }
    }

    override fun deleteFromFavorite(venue: Place): Completable {
        return Completable.fromAction {
            val detailedVenueDbEntity = DetailedVenueMapper.map(venue).venue
            dao.update(detailedVenueDbEntity.copy(isFavorite = false))
        }
    }

}
